pkg/leeway/package.go

```diff
@@ -1002,7 +1006,65 @@ func resolveBuiltinGitVariables(pkg *Package, vars map[string]string) error {
 		return fmt.Errorf("package sources are not within a Git working copy")
 	}
 
+	if len(commit) != 40 {
+		return fmt.Errorf("unexpected commit hash length [%d]: [%s]", len(commit), commit)
+	}
+
+	commitShort := commit[0:7]
+	isDirty := pkg.C != nil && pkg.C.Git() != nil && pkg.C.Git().dirty && isPkgDirty(pkg, pkg.C.Git())
+	if isDirty {
+		version, err := pkg.Version()
+		if err != nil {
+			return fmt.Errorf("unable to get version hash: [%v]", err)
+		}
+
+		commitShort = fmt.Sprintf("%s-%s", commitShort, version)
+		commit = fmt.Sprintf("%s-%s", commit, version)
+	}
+
 	vars[BuildinArgGitCommit] = commit
+	vars[BuildinArgGitCommitShort] = commitShort
 
 	return nil
 }
+
+// isPkgDirty recursively checks whether a package and all its dependencies are dirty
+// by comparing the git working copy with the content of the package and if there is a match, marks the package as dirty
+func isPkgDirty(pkg *Package, info *GitInfo) bool {
+	for _, s := range pkg.Sources {
+		// paths in `GitInfo.dirtyFiles` don't contain the git work dir, so we strip it
+		source := strings.TrimPrefix(s, info.WorkingCopyLoc+string(os.PathSeparator))
+		if _, ok := info.dirtyFiles[source]; ok {
+			log.WithFields(log.Fields{
+				"pkg":  pkg.Name,
+				"file": source,
+			}).Debug("Package is dirty")
+			return true
+		}
+
+		// if we get a dir in our dirty files (dirs), it doesn't have a trailing slash, so we add it
+		dirMatch := filepath.Dir(source) + string(os.PathSeparator)
+		if _, ok := info.dirtyFiles[dirMatch]; ok {
+			log.WithFields(log.Fields{
+				"pkg":  pkg.Name,
+				"dir":  dirMatch,
+				"file": source,
+			}).Debug("Package is dirty")
+			return true
+		}
+	}
+
+	for _, d := range pkg.GetTransitiveDependencies() {
```

Comment:
Since `GetTransitiveDependencies()` returns not only all direct dependencies of this package but also all indirect dependencies, it's redundant to `isPkgDirty()` beeing recursive. Proposal:
Either call `GetTransitiveDependencies()` and don't recurse, or call `pkg.GetDependencies()` and keep the recursion.